Delta stepping

SChudakov opened the pull request:
Implemented delta-stepping algorithm discussed in  #663

[1] U. Meyer, P. Sanders, Δ-stepping: a parallelizable shortest path algorithm, Journal of Algorithms, Volume 49, Issue 1, 2003, Pages 114-152, ISSN 0196-6774, https://doi.org/10.1016/S0196-6774(03)00076-2.
  
<!-- describe the changes you have made here: what, why, …
     Link issues by using the following pattern: [#333](https://github.com/jgrapht/jgrapht/issues/333).
     The title of the PR must not reference an issue, because GitHub does not support autolinking there. -->

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,366 @@
+     */
+    private Map<V, Set<E>> light;
+    /**
+     * Map with light edges for each vertex. An edge is
```
"light"->"heavy"

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,366 @@
+
+    /**
+     * Constructs a new instance of the algorithm for a given graph.
+     * Initializes {@link #delta} to $0.0$ to preserve lazy computation stype.
```
typo

d-michail commented on jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPathTest.java:
```diff
@@ -0,0 +1,142 @@
+ *
+ * @author Semen Chudakov
+ */
+public class DeltaSteppingShortestPathTest {
```
Could you add a test which creates a few random graphs, perhaps using the gnp or the Barabasi-Albert generator, executes your implementation and Dijkstra's algorithm and asserts that the shortest path distances are the same? We have similar tests in the library.

jbduncan commented:
> It seems that with a dedicated ForkJoinPool, you can give a user more control over how many threads are used?

@jkinable As someone observing JGraphT from the outside, I very much agree with this!

jbduncan commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,366 @@
+     * @return max edge weight
+     */
+    private Optional<Double> maxEdgeWeight() {
+        return graph.edgeSet().stream().map(graph::getEdgeWeight).max(Double::compare);
```
Indeed. It can be eliminated by replacing `.max(Double::compare)` with `.mapToDouble(d -> d).max()`. However this then requires that the result type be `OptionalDouble`, which takes less space but has less methods.

jbduncan commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,366 @@
+    private double findDelta() {
+        Optional<Double> maxEdgeWeight = maxEdgeWeight();
+        if (maxEdgeWeight.isPresent()) {
+            int maxOutDegree = graph.vertexSet().stream().map(graph::outDegreeOf).max(Integer::compare).orElse(0);
```
Consider replacing `.max(Integer::compare)` with `.mapToInt(i -> i).max()`.

jbduncan commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,366 @@
+        return verticesDataMap.values().stream()
+                .map(triple -> triple.get().getFirst())
+                .filter(bucketIndex -> bucketIndex >= 0)
+                .min(Integer::compare).orElse(-1);
```
Consider replacing `.min(Integer::compare)` with `.mapToInt(i -> i).min()`.

jbduncan commented:
> > It seems that with a dedicated ForkJoinPool, you can give a user more control over how many threads are used?
> 
> @jkinable As someone observing JGraphT from the outside, I very much agree with this!

I'll add that this may be desirable if e.g. one wants to have one thread pool for CPU-heavy operations and another for IO-heavy operations, or if one is working within an application container.

jkinable commented:
So if I interpret these results correctly, then our Single threaded Dijkstra is outperforming Delta stepping with 8 Threads? Are there any obvious performance improvements you could think of?

SChudakov commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,371 @@
+     */
+    private int numOfBuckets() {
+        return getMaxEdgeWeight().map(maxWeight -> (int) (Math.ceil(maxWeight / delta) + 1)).orElse(1);
+    }
```
To my mind, it is not necessary to changed  `getMaxEdgeWeight()` result type to `OptionalDouble` , because it will result in more sophisticated code here, since `OptionalDouble`  does not have `map()` method.

jbduncan replied:
Okay, that sounds perfectly reasonable to me! I consider `map()` to be an important method, so I'm happy for `Optional<Double>` to be returned here.

SChudakov commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,371 @@
+ * algorithm is fully sequential and the {@link BellmanFordShortestPath} algorithm
+ * has high parallelism since all edges can be relaxed in parallel, the delta-stepping
+ * introduces parameter delta, which, when chooses optimally, yields still good parallelism
+ * and at the same time enables avoiding too many re-relaxations of the edges.
```
Also I have not metioned any parallilisation details here yet, because I  will make changes to it in order to improve performance.

jkinable replied:
ok sounds good. Forgot to mention: please include runtime and space complexity information here using big-O notation.

jkinable commented on jgrapht-core/src/test/java/org/jgrapht/perf/shortestpath/DeltaSteppingShortestPathPerformance.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Benchmark
+    public ShortestPathAlgorithm.SingleSourcePaths<Integer, DefaultWeightedEdge> testDense(DenseGraphData data) {
```
rename these methods to testDenseDeltaStepping. That way the name suggests which algorithm is used.
testSparse-> testSparseDeltaStepping

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,384 @@
+     * @return max edge weight
+     */
+    private Optional<Double> getMaxEdgeWeight() {
+        return graph.edgeSet().stream().map(graph::getEdgeWeight).max(Double::compare);
```
You can also perform this in parallel, using `edgeSet().parallelStream()`.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,384 @@
+     * Initializes {@link #light}, {@link #heavy} and {@link #verticesDataMap} fields.
+     */
+    private void init() {
+        light = new HashMap<>();
```
Can you change this to `ConcurrentHashMap` and perform the write in parallel?

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,384 @@
+    private double findDelta() {
+        Optional<Double> maxEdgeWeight = getMaxEdgeWeight();
+        if (maxEdgeWeight.isPresent()) {
+            int maxOutDegree = graph.vertexSet().stream().mapToInt(graph::outDegreeOf).max().orElse(0);
```
This can also happen in parallel.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,384 @@
+     * Fills {@link #light}, {@link #heavy} and{@link #verticesDataMap} fields.
+     */
+    private void fillMaps() {
+        graph.vertexSet().forEach(v -> {
```
If you change the maps to `ConcurrentHashMap` can you perform this also in parallel?

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:
```diff
@@ -0,0 +1,384 @@
+     * to avoid threads synchronisation. Thus a thread can safely update
+     * the information using standard CAS function.
+     */
+    private Map<V, AtomicReference<Triple<Integer, Double, E>>> verticesDataMap;
```
I am not really sure whether this representation is enough in order to get good performance. You are filtering all vertices in order to find the vertices inside some bucket. I guess that you need to explicitly maintain the buckets.

SChudakov commented:
i`ve made some changes to the algorithm, so now it clearly outperfors Dijkstra on random graphs. The benchmark results with 12 threads are following:

| Benchmark               | (edgeDegree)  | (graphSize) | (numOfVertices)| Mode | Cnt  |   Score |      Error | Units  |
|-------------------------|---------------|-------------|----------------|------|------|---------|------------|--------|
| testBellmanFordDense    |          N/A  |       1000  |            N/A | avgt |   8  | 248.876 |  ±  94.379 | ms/op |
| testBellmanFordRandom   |          100  |        N/A  |           5000 | avgt |   8  | 397.204 |  ±  86.428 | ms/op |
| testBellmanFordSparse   |           50  |      10000  |            N/A | avgt |   8  |1076.279 |  ± 628.313 | ms/op |
| testDeltaSteppingDense  |          N/A  |       1000  |            N/A | avgt |   8  |  41.869 |  ±   4.885 | ms/op |
| testDeltaSteppingRandom |          100  |        N/A  |           5000 | avgt |   8  |  53.476 |  ±   1.865 | ms/op |
| testDeltaSteppingSparse |           50  |      10000  |            N/A | avgt |   8  |  95.734 |  ±   8.119 | ms/op |
| testDijkstraDense       |          N/A  |       1000  |            N/A | avgt |   8  |  36.249 |  ±   8.407 | ms/op |
| testDijkstraRandom      |          100  |        N/A  |           5000 | avgt |   8  |  79.489 |  ±   8.893 | ms/op |
| testDijkstraSparse      |           50  |      10000  |            N/A | avgt |   8  |  34.427 |  ±   4.214 | ms/op |

SChudakov commented:
I`ve made some changes to the algorithm, so now it clearly outperfors Dijkstra on random graphs. The benchmark results with 12 threads are following:

| Benchmark               | (edgeDegree)  | (graphSize) | (numOfVertices)| Mode | Cnt  |   Score |      Error | Units  |
|-------------------------|---------------|-------------|----------------|------|------|---------|------------|--------|
| testBellmanFordDense    |          N/A  |       1000  |            N/A | avgt |   8  | 248.876 |  ±  94.379 | ms/op |
| testBellmanFordRandom   |          100  |        N/A  |           5000 | avgt |   8  | 397.204 |  ±  86.428 | ms/op |
| testBellmanFordSparse   |           50  |      10000  |            N/A | avgt |   8  |1076.279 |  ± 628.313 | ms/op |
| testDeltaSteppingDense  |          N/A  |       1000  |            N/A | avgt |   8  |  41.869 |  ±   4.885 | ms/op |
| testDeltaSteppingRandom |          100  |        N/A  |           5000 | avgt |   8  |  53.476 |  ±   1.865 | ms/op |
| testDeltaSteppingSparse |           50  |      10000  |            N/A | avgt |   8  |  95.734 |  ±   8.119 | ms/op |
| testDijkstraDense       |          N/A  |       1000  |            N/A | avgt |   8  |  36.249 |  ±   8.407 | ms/op |
| testDijkstraRandom      |          100  |        N/A  |           5000 | avgt |   8  |  79.489 |  ±   8.893 | ms/op |
| testDijkstraSparse      |           50  |      10000  |            N/A | avgt |   8  |  34.427 |  ±   4.214 | ms/op |

jkinable commented:
What's the difference between graphsize and numOfVertices? Also, why are some of them N/A?
The performance still seems quite poor. With 12 threads Delta stepping still has a hard time to outperform Dijkstra. Are there any other performance improvements you could think of?

SChudakov commented:
There is non difference between graphSize and numOfVerices, that`s just a divergence in variable naming. I will fix it. A test parameter is set to N/A whenever it is not generated for a test. Than happens for example for dense graphs test, because the edgeDegree which is present in sparse and random graphs tests, is not needed. 
Ok, I will consider some further improvements.

d-michail commented:
I still think that the bucket filtering that you are performing could be improved.

SChudakov commented:
Thanks! I will consider that too.

jkinable commented:
Any progress on this one?

SChudakov commented:
I am currently developing the idea proposed by @d-michail about maintaining explicitly the bucket structure. My current progress is available in `ds_skip-list-set` branch  on my local fork. In brief, I am finished with implementation of the bucket structure and I am working on a synchronization bug in the relaxation procedure. As soon as I am finished with that, I will merge all changes in this branch.

SChudakov commented:
I`m finished with fixing the bug so now I`m ready to present the changes I`ve made so far. The following points are changed: 

- the bucket structure is implemented

- the `light`, `heavy` and `distanceAndPredecessorMap` are now filled in a fully concurrent way

- the `loadBalancing` parameter is added to avoid unnecessary computation for random graphs

- the constructors are simplified (in particular the `init()` methods is removed)

Benchmarks result are in the next comment

SChudakov commented:
In general the algorithm became slightly faster. The notable thing here is that the increasing of the number of vertices in the dense graph benchmark does not slow the execution time as is did with the previous implementation.

| Benchmark             |(edgeDegree) |(graphSize)|  Mode | Cnt  |   Score |      Error | Units  |
|-----------------------|-------------|-----------|-------|------|---------|------------|--------|
|testDeltaSteppingDense |      N/A    |      1000 |  avgt |  8   |   39.017| ±   3.901  | ms/op  |
|testDeltaSteppingDense |      N/A    |      2000 |  avgt |  8   |  217.967| ±  61.486  | ms/op  |
|testDeltaSteppingDense |      N/A    |      3000 |  avgt |  8   |  446.584| ± 250.533  | ms/op  |
|testDeltaSteppingRandom|       50    |     10000 |  avgt |  8   |   61.858| ±  10.126  | ms/op  |
|testDeltaSteppingRandom|      500    |     10000 |  avgt |  8   |  671.888| ± 543.753  | ms/op  |
|testDijkstraDense      |      N/A    |      1000 |  avgt |  8   |   37.855| ±   9.253  | ms/op  |
|testDijkstraDense      |      N/A    |      2000 |  avgt |  8   |  138.656| ±  13.546  | ms/op  |
|testDijkstraDense      |      N/A    |      3000 |  avgt |  8   |  327.514| ±  16.288  | ms/op  |
|testDijkstraRandom     |       50    |     10000 |  avgt |  8   |   82.280| ±   9.089  | ms/op  |
|testDijkstraRandom     |      500    |     10000 |  avgt |  8   |  950.812| ±  62.004  | ms/op  |
|testBellmanFordDense   |      N/A    |      1000 |  avgt |  8   |  208.224| ±  65.462  | ms/op  |
|testBellmanFordDense   |      N/A    |      2000 |  avgt |  8   | 1001.008| ± 195.966  | ms/op  |
|testBellmanFordDense   |      N/A    |      3000 |  avgt |  8   | 2297.833| ± 221.676  | ms/op  |
|testBellmanFordRandom  |       50    |     10000 |  avgt |  8   |  422.074| ±  57.771  | ms/op  |
|testBellmanFordRandom  |      500    |     10000 |  avgt |  8   | 4254.418| ± 383.548  | ms/op  |

SChudakov commented:
I am finished with fixing the bug so now I am ready to present the changes I have made so far. The following points are changed: 

- the bucket structure is implemented

- the `light`, `heavy` and `distanceAndPredecessorMap` are now filled in a fully concurrent way

- the `loadBalancing` parameter is added to avoid unnecessary computation for random graphs

- the constructors are simplified (in particular the `init()` methods is removed)

- the parse graphs in the benchmarks are now also generated randomly

Benchmarks result are in the next comment

SChudakov commented:
In general the algorithm became slightly faster. The notable thing here is that the increasing of the number of vertices in the dense graph benchmark does not slow the execution time as is did with the previous implementation.

| Benchmark             |(edgeDegree) |(graphSize)|  Mode | Cnt  |   Score |      Error | Units  |
|-----------------------|-------------|-----------|-------|------|---------|------------|--------|
|testDeltaSteppingDense |      N/A    |      1000 |  avgt |  8   |   39.017| ±   3.901  | ms/op  |
|testDeltaSteppingDense |      N/A    |      2000 |  avgt |  8   |  217.967| ±  61.486  | ms/op  |
|testDeltaSteppingDense |      N/A    |      3000 |  avgt |  8   |  446.584| ± 250.533  | ms/op  |
|testDeltaSteppingRandom|       50    |     10000 |  avgt |  8   |   61.858| ±  10.126  | ms/op  |
|testDeltaSteppingRandom|      500    |     10000 |  avgt |  8   |  671.888| ± 543.753  | ms/op  |
|testDijkstraDense      |      N/A    |      1000 |  avgt |  8   |   37.855| ±   9.253  | ms/op  |
|testDijkstraDense      |      N/A    |      2000 |  avgt |  8   |  138.656| ±  13.546  | ms/op  |
|testDijkstraDense      |      N/A    |      3000 |  avgt |  8   |  327.514| ±  16.288  | ms/op  |
|testDijkstraRandom     |       50    |     10000 |  avgt |  8   |   82.280| ±   9.089  | ms/op  |
|testDijkstraRandom     |      500    |     10000 |  avgt |  8   |  950.812| ±  62.004  | ms/op  |
|testBellmanFordDense   |      N/A    |      1000 |  avgt |  8   |  208.224| ±  65.462  | ms/op  |
|testBellmanFordDense   |      N/A    |      2000 |  avgt |  8   | 1001.008| ± 195.966  | ms/op  |
|testBellmanFordDense   |      N/A    |      3000 |  avgt |  8   | 2297.833| ± 221.676  | ms/op  |
|testBellmanFordRandom  |       50    |     10000 |  avgt |  8   |  422.074| ±  57.771  | ms/op  |
|testBellmanFordRandom  |      500    |     10000 |  avgt |  8   | 4254.418| ± 383.548  | ms/op  |

SChudakov commented:
I am now exploring the possibility to further improve the performance by creating own implementation of the buckets instead of using `ConcurrentSkipListSet`. This will potentially enable to do this operation `removed.addAll(bucketElements);` in constant time my simply merging 2 linked lists and then traverse the result to split vertices among the threads.

SChudakov commented:
I am finished with fixing the bug so now I am ready to present the changes I have made so far. The following points are changed: 

- the bucket structure is implemented

- the `light`, `heavy` and `distanceAndPredecessorMap` are now filled in a fully concurrent way

- the `loadBalancing` parameter is added to avoid unnecessary computation for random graphs

- the constructors are simplified (in particular the `init()` methods is removed)

- the parse graphs in the benchmarks are now also generated randomly

Benchmarks result are in the next comment

SChudakov commented:
In general the algorithm became slightly faster. The notable thing here is that the increase of the number of vertices in the dense graph benchmark does not slow the execution time as is did with the previous implementation.

| Benchmark             |(edgeDegree) |(graphSize)|  Mode | Cnt  |   Score |      Error | Units  |
|-----------------------|-------------|-----------|-------|------|---------|------------|--------|
|testDeltaSteppingDense |      N/A    |      1000 |  avgt |  8   |   39.017| ±   3.901  | ms/op  |
|testDeltaSteppingDense |      N/A    |      2000 |  avgt |  8   |  217.967| ±  61.486  | ms/op  |
|testDeltaSteppingDense |      N/A    |      3000 |  avgt |  8   |  446.584| ± 250.533  | ms/op  |
|testDeltaSteppingRandom|       50    |     10000 |  avgt |  8   |   61.858| ±  10.126  | ms/op  |
|testDeltaSteppingRandom|      500    |     10000 |  avgt |  8   |  671.888| ± 543.753  | ms/op  |
|testDijkstraDense      |      N/A    |      1000 |  avgt |  8   |   37.855| ±   9.253  | ms/op  |
|testDijkstraDense      |      N/A    |      2000 |  avgt |  8   |  138.656| ±  13.546  | ms/op  |
|testDijkstraDense      |      N/A    |      3000 |  avgt |  8   |  327.514| ±  16.288  | ms/op  |
|testDijkstraRandom     |       50    |     10000 |  avgt |  8   |   82.280| ±   9.089  | ms/op  |
|testDijkstraRandom     |      500    |     10000 |  avgt |  8   |  950.812| ±  62.004  | ms/op  |
|testBellmanFordDense   |      N/A    |      1000 |  avgt |  8   |  208.224| ±  65.462  | ms/op  |
|testBellmanFordDense   |      N/A    |      2000 |  avgt |  8   | 1001.008| ± 195.966  | ms/op  |
|testBellmanFordDense   |      N/A    |      3000 |  avgt |  8   | 2297.833| ± 221.676  | ms/op  |
|testBellmanFordRandom  |       50    |     10000 |  avgt |  8   |  422.074| ±  57.771  | ms/op  |
|testBellmanFordRandom  |      500    |     10000 |  avgt |  8   | 4254.418| ± 383.548  | ms/op  |

SChudakov commented:
I am now exploring the possibility to further improve the performance by creating own implementation of the buckets instead of using `ConcurrentSkipListSet`. This will potentially enable to do this operation `removed.addAll(bucketElements);` in constant time my simply merging 2 linked lists and then traverse the result to split vertices among the threads.

SChudakov commented:
Changes made so far:

- all working threads now receive vertices from a general queue
- developed separates procedure to handle relaxation light and heavy relax requests
- the amount of operations between algorithm iterations is reduced to only adding vertices to the queue 
- setup level in  benchmarks is changed to Level.Trial

The next step I am going to take is to explore the C++ implementation of the delta-stepping algorithm in the Parallel Boost Graph Library [here](https://www.boost.org/doc/libs/1_66_0/libs/graph_parallel/doc/html/dijkstra_shortest_paths.html#delta-stepping-algorithm).

jkinable commented:
Very nice, thanks. Do you have an idea as to how your modifications impact the performance? It would also be good to have a look at the Parallel Boost Graph Library to understand how their parallel implementation compares against their sequential implementation so understand how much we could potentially gain from the parallellization.

SChudakov commented:
Just forgot to provide the newest benchmarks results. Here they are:

| Benchmark             |(edgeDegree) |(graphSize)|  Mode | Cnt  |   Score |      Error | Units  |
|-----------------------|-------------|-----------|-------|------|---------|------------|--------|
|testBellmanFordDense    |          N/A|    1000   | avgt  |   8  |  189.664| ±   4.194  | ms/op |
|testBellmanFordDense    |          N/A|    2000   | avgt  |   8  | 1113.890| ±  13.835  | ms/op |
|testBellmanFordDense    |          N/A|    3000   | avgt  |   8  | 2309.379| ± 332.512  | ms/op |
|testBellmanFordRandom   |           50|   10000   | avgt  |   8  |  464.139| ±  18.248  | ms/op |
|testBellmanFordRandom   |          500|   10000   | avgt  |   8  | 4978.569| ± 221.840  | ms/op |
|testDeltaSteppingDense  |          N/A|    1000   | avgt  |   8  |   35.166| ±   0.411  | ms/op |
|testDeltaSteppingDense  |          N/A|    2000   | avgt  |   8  |  146.303| ±   1.420  | ms/op |
|testDeltaSteppingDense  |          N/A|    3000   | avgt  |   8  |  354.995| ±   2.377  | ms/op |
|testDeltaSteppingRandom |           50|   10000   | avgt  |   8  |   57.146| ±   1.508  | ms/op |
|testDeltaSteppingRandom |          500|   10000   | avgt  |   8  |  538.635| ±  22.334  | ms/op |
|testDijkstraDense       |          N/A|    1000   | avgt  |   8  |   33.482| ±   1.133  | ms/op |
|testDijkstraDense       |          N/A|    2000   | avgt  |   8  |  141.431| ±   2.247  | ms/op |
|testDijkstraDense       |          N/A|    3000   | avgt  |   8  |  336.488| ±  62.677  | ms/op |
|testDijkstraRandom      |           50|   10000   | avgt  |   8  |   83.938| ±   1.571  | ms/op |
|testDijkstraRandom      |          500|   10000   | avgt  |   8  |  823.150| ±  21.696  | ms/op |


The main boost in delta-stepping performance was provided by changing the setup level in benchmarks.

SChudakov commented:
Changes I have made this time:
- removed dividing edges into light and heavy;
- check for negative edge weights as well as max edge weight computation is now performed using RecursiveTask, which is submitted to the ForkJoinPool;
- removed BucketStructure class as well as other redundant methods, which are used only once;

Benchmarks results:

| Benchmark             |(edgeDegree) |(graphSize)|  Mode | Cnt  |   Score |      Error | Units  |
|-----------------------|-------------|-----------|-------|------|---------|------------|--------|
|testDeltaSteppingDense |          N/A|     1000  | avgt  |    8 |  23.321 |  ± 0.437   |  ms/op |
|testDeltaSteppingDense |          N/A|     2000  | avgt  |    8 |  92.171 |  ± 0.976   |  ms/op |
|testDeltaSteppingDense |          N/A|     3000  | avgt  |    8 | 196.248 |  ± 1.626   |  ms/op |
|testDeltaSteppingRandom|           50|    10000  | avgt  |    8 |  44.966 |  ± 0.325   |  ms/op |
|testDeltaSteppingRandom|          500|    10000  | avgt  |    8 | 362.491 |  ± 2.968   |  ms/op |
|testBellmanFordDense   |          N/A|    1000   | avgt  |   8  |  189.664| ±   4.194  | ms/op |
|testBellmanFordDense   |          N/A|    2000   | avgt  |   8  | 1113.890| ±  13.835  | ms/op |
|testBellmanFordDense   |          N/A|    3000   | avgt  |   8  | 2309.379| ± 332.512  | ms/op |
|testBellmanFordRandom  |           50|   10000   | avgt  |   8  |  464.139| ±  18.248  | ms/op |
|testBellmanFordRandom  |          500|   10000   | avgt  |   8  | 4978.569| ± 221.840  | ms/op |
|testDijkstraDense      |          N/A|    1000   | avgt  |   8  |   33.482| ±   1.133  | ms/op |
|testDijkstraDense      |          N/A|    2000   | avgt  |   8  |  141.431| ±   2.247  | ms/op |
|testDijkstraDense      |          N/A|    3000   | avgt  |   8  |  336.488| ±  62.677  | ms/op |
|testDijkstraRandom     |           50|   10000   | avgt  |   8  |   83.938| ±   1.571  | ms/op |
|testDijkstraRandom     |          500|   10000   | avgt  |   8  |  823.150| ±  21.696  | ms/op |

The gap in performance tends to increase with the increasing of the graph size.

SChudakov commented:
I have added benchmarks for other types of random graphs generators, such as `GnpRandomGraphGenerator`, `BarabasiAlbertGraphGenerator` and `WattsStrogatzGraphGenerator`.
 Here are the results:

|Benchmark                      | (k) | (m) | (m0) | (numOfVertices)   | (p)  | Mode |Cnt   | Score     | Error   | Units |
|-------------------------------|-----|-----|------|-------------------|------|------|------|-----------|---------|-------|
|testBellmanFordBarabasiAlbert  |  N/A|   50|  1000|     10000         |  N/A | avgt | 8    |  414.016  |±   4.458|  ms/op|
|testBellmanFordBarabasiAlbert  |  N/A|  500|  1000|     10000         |  N/A | avgt | 8    | 2809.368  |± 337.355|  ms/op|
|testBellmanFordGnp             |  N/A|  N/A|   N/A|     10000         | 0.01 | avgt | 8    |  297.056  |±   6.066|  ms/op|
|testBellmanFordGnp             |  N/A|  N/A|   N/A|     10000         | 0.05 | avgt | 8    | 1508.716  |±  37.265|  ms/op|
|testBellmanFordWattsStogatz    |  100|  N/A|   N/A|     10000         | 0.05 | avgt | 8    |  204.976  |±   2.521|  ms/op|
|testBellmanFordWattsStogatz    |  100|  N/A|   N/A|     10000         |  0.5 | avgt | 8    |  285.564  |±   2.582|  ms/op|
|testBellmanFordWattsStogatz    | 1000|  N/A|   N/A|     10000         | 0.05 | avgt | 8    | 2656.451  |±  37.423|  ms/op|
|testBellmanFordWattsStogatz    | 1000|  N/A|   N/A|     10000         |  0.5 | avgt | 8    | 3620.227  |±  71.072|  ms/op|
|testDeltaSteppingBarabasiAlbert|  N/A|   50|  1000|     10000         |  N/A | avgt | 8    |   57.078  |±   0.394|  ms/op|
|testDeltaSteppingBarabasiAlbert|  N/A|  500|  1000|     10000         |  N/A | avgt | 8    |  277.189  |±   3.735|  ms/op|
|testDeltaSteppingGnp           |  N/A|  N/A|   N/A|     10000         | 0.01 | avgt | 8    |   33.578  |±   0.188|  ms/op|
|testDeltaSteppingGnp           |  N/A|  N/A|   N/A|     10000         | 0.05 | avgt | 8    |  139.365  |±   0.734|  ms/op|
|testDeltaSteppingWattsStogatz  |  100|  N/A|   N/A|     10000         | 0.05 | avgt | 8    |   33.945  |±   0.449|  ms/op|
|testDeltaSteppingWattsStogatz  |  100|  N/A|   N/A|     10000         |  0.5 | avgt | 8    |   35.429  |±   0.663|  ms/op|
|testDeltaSteppingWattsStogatz  | 1000|  N/A|   N/A|     10000         | 0.05 | avgt | 8    |  219.756  |±   2.519|  ms/op|
|testDeltaSteppingWattsStogatz  | 1000|  N/A|   N/A|     10000         |  0.5 | avgt | 8    |  300.901  |±   2.604|  ms/op|
|testDijkstraBarabasiAlbert     |  N/A|   50|  1000|     10000         |  N/A | avgt | 8    |   87.814  |±   1.781|  ms/op|
|testDijkstraBarabasiAlbert     |  N/A|  500|  1000|     10000         |  N/A | avgt | 8    |  481.330  |±   8.087|  ms/op|
|testDijkstraGnp                |  N/A|  N/A|   N/A|     10000         | 0.01 | avgt | 8    |   59.397  |±   1.008|  ms/op|
|testDijkstraGnp                |  N/A|  N/A|   N/A|     10000         | 0.05 | avgt | 8    |  263.858  |±   3.532|  ms/op|
|testDijkstraWattsStogatz       |  100|  N/A|   N/A|     10000         | 0.05 | avgt | 8    |   49.221  |±   0.514|  ms/op|
|testDijkstraWattsStogatz       |  100|  N/A|   N/A|     10000         |  0.5 | avgt | 8    |   65.936  |±   0.618|  ms/op|
|testDijkstraWattsStogatz       | 1000|  N/A|   N/A|     10000         | 0.05 | avgt | 8    |  392.955  |±   7.903|  ms/op|
|testDijkstraWattsStogatz       | 1000|  N/A|   N/A|     10000         |  0.5 | avgt | 8    |  557.114  |±  26.769|  ms/op|
What do you think about it?

SChudakov commented:
I have made some final refinements to this PR.

jkinable commented:
@SChudakov the computational experiments, with how many threads did you perform them?

@d-michail you left some comments as well. Could you check whether everything looks fine now?

This PR add a DeltaStepping performance test. There already exists a `DijkstraShortestPathPerformanceTest`. Would it make sense to merge them?

After the PR has been merged, we would need to improve the description (javadoc) a bit.

SChudakov commented:
I have done the performance tests with 12 threads.

SChudakov commented:
I have done performance comparison with the Boost implementation of the delta stepping algorithm.

SChudakov commented:
Dijkstra`s algorithm java


----------------------------------------------------------
|Benchmark      |          Parameters   |   Score | Units|
|---------------|-----------------------|---------|------|
|BarabasiAlbert |m0=1000 n=10000 m=50   | 190.530 | ms/op|
|BarabasiAlbert |m0=1000 n=10000 m=500  |1007.620 | ms/op|
|Complete       |n=1000                 |  71.960 | ms/op|
|Complete       |n=2000                 | 320.237 | ms/op|
|Complete       |n=3000                 | 740.949 | ms/op|
|Gnm            |n=10000 m=50           | 154.922 | ms/op|
|Gnm            |n=10000 m=500          |1479.639 | ms/op|
|Gnp            |n=10000 p=0.01         | 123.218 | ms/op|
|Gnp            |n=10000 p=0.05         | 561.838 | ms/op|
|WattsStogatz   |n=10000 k=100   p=0.05 |  94.081 | ms/op|
|WattsStogatz   |n=10000 k=100   p=0.5  | 125.495 | ms/op|
|WattsStogatz   |n=10000 k=1000  p=0.05 | 941.069 | ms/op|
|WattsStogatz   |n=10000 k=1000  p=0.5  |1199.247 | ms/op|
----------------------------------------------------------

SChudakov commented:
Delta stepping algorithm java

threads = 4

-----------------------------------------------------------
|Benchmark      |          Parameters   |   Score  | Units|
|---------------|-----------------------|----------|------|
|BarabasiAlbert | m0=1000 n=10000 m=50  | 144.741  | ms/op|
|BarabasiAlbert | m0=1000 n=10000 m=500 | 504.503  | ms/op|
|Complete       | n=1000                |  78.760  | ms/op|
|Complete       | n=2000                | 278.605  | ms/op|
|Complete       | n=3000                | 612.809  | ms/op|
|Gnm            | n=10000 m=50          | 123.566  | ms/op|
|Gnm            | n=10000 m=500         |1053.584  | ms/op|
|Gnp            | n=10000 p=0.01        | 110.163  | ms/op|
|Gnp            | n=10000 p=0.05        | 413.752  | ms/op|
|WattsStogatz   | n=10000 k=100   p=0.05|  78.892  | ms/op|
|WattsStogatz   | n=10000 k=100   p=0.5 |  86.153  | ms/op|
|WattsStogatz   | n=10000 k=1000  p=0.05| 492.542  | ms/op|
|WattsStogatz   | n=10000 k=1000  p=0.5 | 674.381  | ms/op|
-----------------------------------------------------------

threads = 8

----------------------------------------------------------
|Benchmark      |          Parameters   |   Score | Units|
|---------------|-----------------------|---------|------|
|BarabasiAlbert |m0=1000 n=10000 m=50   | 155.172 | ms/op|
|BarabasiAlbert |m0=1000 n=10000 m=500  |  527.23 | ms/op|
|Complete       |n=1000                 |  57.884 | ms/op|
|Complete       |n=2000                 | 193.327 | ms/op|
|Complete       |n=3000                 | 440.418 | ms/op|
|Gnm            |n=10000 m=50           |  84.501 | ms/op|
|Gnm            |n=10000 m=500          | 642.653 | ms/op|
|Gnp            |n=10000 p=0.01         |  80.818 | ms/op|
|Gnp            |n=10000 p=0.05         | 285.254 | ms/op|
|WattsStogatz   |n=10000 k=100   p=0.05 |  80.216 | ms/op|
|WattsStogatz   |n=10000 k=100   p=0.5  |  83.125 | ms/op|
|WattsStogatz   |n=10000 k=1000  p=0.05 | 437.280 | ms/op|
|WattsStogatz   |n=10000 k=1000  p=0.5  | 743.430 | ms/op|
----------------------------------------------------------

threads = 12

----------------------------------------------------------
|Benchmark      |          Parameters   |   Score | Units|
|---------------|-----------------------|---------|------|
|BarabasiAlbert |m0=1000 n=10000 m=50   |  136.915| ms/op|
|BarabasiAlbert |m0=1000 n=10000 m=500  |  411.570| ms/op|
|Complete       |n=1000                 |   50.839| ms/op|
|Complete       |n=2000                 |  159.700| ms/op|
|Complete       |n=3000                 |  354.834| ms/op|
|Gnm            |n=10000 m=50           |   72.604| ms/op|
|Gnm            |n=10000 m=500          |  509.083| ms/op|
|Gnp            |n=10000 p=0.01         |   71.379| ms/op|
|Gnp            |n=10000 p=0.05         |  230.132| ms/op|
|WattsStogatz   |n=10000 k=100   p=0.05 |   69.438| ms/op|
|WattsStogatz   |n=10000 k=100   p=0.5  |   71.118| ms/op|
|WattsStogatz   |n=10000 k=1000  p=0.05 |  357.472| ms/op|
|WattsStogatz   |n=10000 k=1000  p=0.5  |  512.478| ms/op|
----------------------------------------------------------

SChudakov commented:
Dijkstra`s algorithm cpp

-----------------------------------------------------------------
|Benchmark      |       Parameters     |        Time     | Units|
|---------------|----------------------|-----------------|------|
|BarabasiAlbert |m0=1000 n=10000 m=50  |     5.34 ms     | ms/op|
|BarabasiAlbert |m0=1000 n=10000 m=500 |     5.05 ms     | ms/op|
|comlete        |n=1000                |     4.85 ms     | ms/op|
|comlete        |n=2000                |     17.2 ms     | ms/op|
|comlete        |n=3000                |     34.9 ms     | ms/op|
|Gnm            |n=10000 m=500         |     67.0 ms     | ms/op|
|Gnm            |n=10000 m=50          |     15.1 ms     | ms/op|
|Gnp            |n=10000 p=0.01        |     12.7 ms     | ms/op|
|Gnp            |n=10000 p=0.05        |     32.4 ms     | ms/op|
|WattsStogatz   |n=10000 k=100 p=0.05  |     14.9 ms     | ms/op|
|WattsStogatz   |n=10000 k=100 p=0.5   |     15.1 ms     | ms/op|
|WattsStogatz   |n=10000 k=1000 k=0.05 |     61.1 ms     | ms/op|
|WattsStogatz   |n=10000 k=1000 k=0.5  |     63.1 ms     | ms/op|
-----------------------------------------------------------------

SChudakov commented:
Delta stepping algorithm C++

threads = 4

------------------------------------------------------------------
|Benchmark      |       Parameters      |        Time     | Units|
|---------------|-----------------------|-----------------|------|
|BarabasiAlbert |m0=1000 n=10000 m=50   |          335.9  | ms/op|
|BarabasiAlbert |m0=1000 n=10000 m=500  |          1624.8 | ms/op|
|comlete        |n=1000                 |          74.9   | ms/op|
|comlete        |n=2000                 |          317.9  | ms/op|
|comlete        |n=3000                 |          790.6  | ms/op|
|Gnm            |n=10000 m=50 0         |          108.1  | ms/op|
|Gnm            |n=10000 m=500 0        |          1055.5 | ms/op|
|Gnp            |n=10000 p=0.01 0       |          104.3  | ms/op|
|Gnp            |n=10000 p=0.05 0       |          493.5  | ms/op|
|WattsStogatz   |n=10000 k=100 k=0.05 0 |          68.6   | ms/op|
|WattsStogatz   |n=10000 k=1000 k=0.5 0 |          953.7  | ms/op|
|WattsStogatz   |n=10000 k=1000 k=0.05 0|          913.3  | ms/op|
|WattsStogatz   |n=10000 k=k=0.5 0.5 0  |          95.1   | ms/op|
------------------------------------------------------------------

threads = 8

-------------------------------------------------------------------
|Benchmark      |       Parameters       |        Time     | Units|
|---------------|------------------------|-----------------|------|
|BarabasiAlbert | m0=1000 n=10000 m=50   |       367.1     | ms/op|
|BarabasiAlbert | m0=1000 n=10000 m=500  |       1341.8    | ms/op|
|comlete        | n=1000                 |       56.8      | ms/op|
|comlete        | n=2000                 |       231.4     | ms/op|
|comlete        | n=3000                 |       551.7     | ms/op|
|Gnm            | n=10000 m=50           |       89.7      | ms/op|
|Gnm            | n=10000 m=500          |       760.8     | ms/op|
|Gnp            | n=10000 p=0.01         |       88.5      | ms/op|
|Gnp            | n=10000 p=0.05         |       378.5     | ms/op|
|WattsStogatz   | n=10000 k=100 k=0.05   |       53        | ms/op|
|WattsStogatz   | n=10000 k=1000 k=0.5   |       708       | ms/op|
|WattsStogatz   | n=10000 k=1000 k=0.05  |       704       | ms/op|
|WattsStogatz   | n=10000 k=k=0.5 0.5    |       74.1      | ms/op|
-------------------------------------------------------------------

threads = 12

-------------------------------------------------------------------
|Benchmark      |       Parameters       |        Time     | Units|
|---------------|------------------------|-----------------|------|
|BarabasiAlbert | m0=1000 n=10000 m=50   |        780.5    | ms/op|
|BarabasiAlbert | m0=1000 n=10000 m=500  |        1485.0   | ms/op|
|comlete        | n=1000                 |        107.8    | ms/op|
|comlete        | n=2000                 |        293.1    | ms/op|
|comlete        | n=3000                 |        630.5    | ms/op|
|Gnm            | n=10000 m=50           |        192.1    | ms/op|
|Gnm            | n=10000 m=500          |        922.2    | ms/op|
|Gnp            | n=10000 p=0.01         |        189.6    | ms/op|
|Gnp            | n=10000 p=0.05         |        490.7    | ms/op|
|WattsStogatz   | n=10000 k=100   p=0.05 |        132.4    | ms/op|
|WattsStogatz   | n=10000 k=100   p=0.5  |        150.2    | ms/op|
|WattsStogatz   | n=10000 k=1000  p=0.05 |        795.0    | ms/op|
|WattsStogatz   | n=10000 k=1000  p=0.5  |        789.6    | ms/op|
-------------------------------------------------------------------

SChudakov commented:
Are there any other steps I should take in order for the PR to be merged?

jkinable commented:
This PR is ready to be merged. However, when merged someone must go over the javadoc to fix some language/grammar issues. I"ll probably have time to look into that this weekend, but go ahead if someone has some additional time available this week.

jkinable commented:
Nicely done. Great addition.

@SChudakov I'm not sure how to interpret the last tables you posted. Are these the boost results? What do we learn from these tables? How does the boost implementation compare against the jgrapht implementation?
Also, do you know something about the performance of the algorithm when more threads are added? E.g. what's the performance curve when running with 1, 2, 4, 8, 16, 32 threads?

SChudakov commented:
1. About the boost performance. 
    In boost library the delta-stepping algorithm is implemented in a  parallel setting. This means that for every worker a single "heavyweight" process is created and scheduled to a separate core. Therefore one can only create N workers for processor with N cores. Additionally for every relaxation of edge, which does not belong to the partition of current process, a message is created which then is passed to the messages queue of process responsible for the edge. This together with the fact that these processes are hard to manage results in a worse performance comparing to the sequential  Dijkstra`s algorithm provided in the library. I have also notices while doing the benchmarks that having other apps started and not necessarily active during the benchmarks execution results in a significant overhead (around 30%) due to the processes being resubmitted to the cores . This does not happened  for the this implementation. 
   The goal of posting the results here was to show that in the situation, when there is one multi-core processor, the concurrent implementation is much preferable. 
   Concluding I can say that both implementation are good but for different kinds of problems. The boost implementation is meant to be used for cluster computations when the graph itself does not fit into RAM of a single machine. The implementation provided here is good to take advantage of the multiple cores and speed up computations for not very large graph instances.

2. The benchmarks.
    I made the performance tests on a 24 cores machine running Ubuntu 16.04. I performed the evaluation on 20 cores maximum (20 and 40 workers correspondingly). This resulted into an overhead of 90%  for boost implementation (to many partitions for such a small graphs) and 20% for this one. However is increase the amount of workers to 14 i gives some benefit of 5%. So once it is not a great speedup I decided not to post these results to this thread.

The bottom line it that the implementation provided here is a good choice if dealing with a single processor with multiple cores.

jkinable commented:
I've made some changes to the javadoc description of the class. Let me know if I made any mistakes.

SChudakov commented:
Nice! Looks good to me.
